// roblox-ts/out/TSTransformer/nodes/jsx/transformJsxAttributes.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.transformJsxAttributes = void 0;
const luau_ast_1 = __importDefault(require("@roblox-ts/luau-ast"));
const assert_1 = require("../../../Shared/util/assert");
const transformExpression_1 = require("../expressions/transformExpression");
const constants_1 = require("../../util/jsx/constants");
const createRoactIndex_1 = require("../../util/jsx/createRoactIndex");
const pointer_1 = require("../../util/pointer");
const types_1 = require("../../util/types");
const typescript_1 = __importDefault(require("typescript"));
function transformJsxInitializer(state, initializer) {
    if (initializer && typescript_1.default.isJsxExpression(initializer)) {
        initializer = initializer.expression;
    }
    if (initializer) {
        return state.capture(() => (0, transformExpression_1.transformExpression)(state, initializer));
    }
    else {
        return [luau_ast_1.default.bool(true), luau_ast_1.default.list.make()];
    }
}
function createJsxAttributeLoop(state, attributesPtrValue, expression, type) {
    const possiblyUndefined = (0, types_1.isPossiblyType)(type, types_1.isUndefinedType);
    if (possiblyUndefined) {
        expression = state.pushToVarIfComplex(expression, "attribute");
    }
    const keyId = luau_ast_1.default.tempId("k");
    const valueId = luau_ast_1.default.tempId("v");
    let statement = luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.ForStatement, {
        ids: luau_ast_1.default.list.make(keyId, valueId),
        expression,
        statements: luau_ast_1.default.list.make(luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.Assignment, {
            left: luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.ComputedIndexExpression, {
                expression: attributesPtrValue,
                index: keyId,
            }),
            operator: "=",
            right: valueId,
        })),
    });
    if (possiblyUndefined) {
        statement = luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.IfStatement, {
            condition: expression,
            statements: luau_ast_1.default.list.make(statement),
            elseBody: luau_ast_1.default.list.make(),
        });
    }
    return statement;
}
function isFlatObject(expression) {
    for (const property of expression.properties) {
        if (!typescript_1.default.isPropertyAssignment(property) || !typescript_1.default.isIdentifier(property.name)) {
            return false;
        }
    }
    return true;
}
function transformSpecialAttribute(state, attribute, attributesPtr) {
    (0, assert_1.assert)(attribute.initializer && typescript_1.default.isJsxExpression(attribute.initializer) && attribute.initializer.expression);
    const expression = attribute.initializer.expression;
    if (typescript_1.default.isObjectLiteralExpression(expression) && isFlatObject(expression)) {
        for (const property of expression.properties) {
            (0, assert_1.assert)(typescript_1.default.isPropertyAssignment(property) && typescript_1.default.isIdentifier(property.name));
            const [init, initPrereqs] = transformJsxInitializer(state, property.initializer);
            if (!luau_ast_1.default.list.isEmpty(initPrereqs)) {
                (0, pointer_1.disableMapInline)(state, attributesPtr);
            }
            state.prereqList(initPrereqs);
            (0, pointer_1.assignToMapPointer)(state, attributesPtr, (0, createRoactIndex_1.createRoactIndex)(attribute.name.text, property.name.text), init);
        }
    }
    else {
        (0, pointer_1.disableMapInline)(state, attributesPtr);
        const init = (0, transformExpression_1.transformExpression)(state, expression);
        const keyId = luau_ast_1.default.tempId("k");
        const valueId = luau_ast_1.default.tempId("v");
        state.prereq(luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.ForStatement, {
            ids: luau_ast_1.default.list.make(keyId, valueId),
            expression: init,
            statements: luau_ast_1.default.list.make(luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.Assignment, {
                left: luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.ComputedIndexExpression, {
                    expression: attributesPtr.value,
                    index: luau_ast_1.default.create(luau_ast_1.default.SyntaxKind.ComputedIndexExpression, {
                        expression: (0, createRoactIndex_1.createRoactIndex)(attribute.name.text),
                        index: keyId,
                    }),
                }),
                operator: "=",
                right: valueId,
            })),
        }));
    }
}
function isSpecialAttribute(state, attribute) {
    (0, assert_1.assert)(state.services.roactSymbolManager);
    const contextualType = state.typeChecker.getContextualType(attribute.parent);
    if (contextualType) {
        const symbol = contextualType.getProperty(attribute.name.text);
        if (symbol) {
            const targetSymbol = typescript_1.default.getSymbolTarget(symbol, state.typeChecker);
            if (targetSymbol === state.services.roactSymbolManager.getSymbolOrThrow(constants_1.EVENT_ATTRIBUTE_NAME) ||
                targetSymbol === state.services.roactSymbolManager.getSymbolOrThrow(constants_1.CHANGE_ATTRIBUTE_NAME)) {
                return true;
            }
        }
    }
    return false;
}
function transformJsxAttribute(state, attribute, attributesPtr) {
    const attributeName = attribute.name.text;
    if (attributeName === constants_1.KEY_ATTRIBUTE_NAME)
        return;
    if (isSpecialAttribute(state, attribute)) {
        transformSpecialAttribute(state, attribute, attributesPtr);
        return;
    }
    const [init, initPrereqs] = transformJsxInitializer(state, attribute.initializer);
    if (!luau_ast_1.default.list.isEmpty(initPrereqs)) {
        (0, pointer_1.disableMapInline)(state, attributesPtr);
        state.prereqList(initPrereqs);
    }
    const name = attributeName === constants_1.REF_ATTRIBUTE_NAME ? (0, createRoactIndex_1.createRoactIndex)(constants_1.REF_ATTRIBUTE_NAME) : luau_ast_1.default.string(attributeName);
    (0, pointer_1.assignToMapPointer)(state, attributesPtr, name, init);
}
function transformJsxAttributes(state, attributes, attributesPtr) {
    for (const attribute of attributes.properties) {
        if (typescript_1.default.isJsxAttribute(attribute)) {
            transformJsxAttribute(state, attribute, attributesPtr);
        }
        else {
            (0, pointer_1.disableMapInline)(state, attributesPtr);
            const expression = (0, transformExpression_1.transformExpression)(state, attribute.expression);
            state.prereq(createJsxAttributeLoop(state, attributesPtr.value, expression, state.getType(attribute.expression)));
        }
    }
}
exports.transformJsxAttributes = transformJsxAttributes;
