// eslint-plugin-roblox-ts/out/rules/module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.module = exports.moduleName = void 0;
const rules_1 = require("../util/rules");
exports.moduleName = "module";
exports.module = rules_1.makeRule({
    name: exports.moduleName,
    meta: {
        type: "problem",
        docs: {
            description: "Enforces file is a module",
            category: "Possible Errors",
            recommended: "error",
            requiresTypeChecking: false,
        },
        fixable: "code",
        messages: {
            moduleViolation: "File must contain at least one import or export statement to be a module.",
        },
        schema: [],
    },
    defaultOptions: [],
    create(context) {
        const service = rules_1.getParserServices(context);
        return {
            Program(node) {
                const tsNode = service.esTreeNodeToTSNodeMap.get(node);
                if (!tsNode.isDeclarationFile && tsNode.externalModuleIndicator === undefined) {
                    context.report({
                        node,
                        messageId: "moduleViolation",
                        fix: fixer => fixer.insertTextBefore(node, "export {};\n"),
                    });
                }
            },
        };
    },
});
